crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1101,6 +1118,67 @@ impl<'db> TypeInferenceBuilder<'db> {
             self.infer_definition(parameter);
         }
         self.infer_body(&function.body);
+
+        if let Some(declared_ty) = function
+            .returns
+            .as_deref()
+            .map(|ret| self.file_expression_type(ret))
+        {
+            fn is_suite_empty(suite: &[ast::Stmt]) -> bool {
+                let mut saw_ellipsis = false;
+                for statement in suite {
+                    if let Some(expr) = statement.as_expr_stmt() {
+                        if expr.value.is_ellipsis_literal_expr() {
+                            saw_ellipsis = true;
+                        } else if expr.value.is_string_literal_expr() {
+                            // docstring
+                        } else {
+                            return false;
+                        }
+                    }
+                }
+                saw_ellipsis
+            }
+            if is_suite_empty(&function.body) {
+                return;
+            }
+            let inferred_ty = if self.types.return_types_and_ranges.is_empty() {
+                KnownClass::NoneType.to_instance(self.db())
```

Comment:
I don't think this logic is quite right. Consider a function like this:

```py
def f(cond: bool) -> int:
    if cond:
        return 1
```

This function will not have an empty `return_type_and_ranges`, but it is still possible for control flow to fall off the end of the function (implicit return `None`).

I think the best way to check for this is to query the final state of the use-def map builder to see whether `scope_start_visibility` is `ALWAYS_FALSE`. This will require recording one new field onto `UseDefMap` (we can call it `can_implicit_return: bool` and add a public method `UseDefMap::can_implicit_return()`) based on checking `scope_start_visibility` in `UseDefMapBuilder::finish`.